### DISCUSSION TITLE
Graphics: Stochastic texture sampling to avoid texture tiling

### DISCUSSION THREAD

**Discussion by HybridDog** [SuperTux/supertux#2352] [Ideas]
The [High-Performance By-Example Noise using a Histogram-Preserving Blending Operator](https://eheitzresearch.wordpress.com/722-2/) paper and an additional [technical chapter](https://eheitzresearch.wordpress.com/738-2/) explain a fast algorithm for stochastic texture sampling.
This could be implemented in SuperTux to reduce visible tiling. Here's an example with one of SuperTux' textures:
![2022-11-19-135534_989x683_scrot](https://user-images.githubusercontent.com/3192173/206516509-765aae46-aeb1-432c-b343-48e1b221f5ee.png)
left: tiled texture, right: texture sampled with stochastic texture sampling

**Comment by Semphriss** [comment]
This looks interesting. Would it be feasible with the SDL renderer, or would it be OpenGL-only?

**Reply by HybridDog** [reply]
I don't know the capabilities of the SDL renderer, so I cannot give a clear answer. The last step of the method is a histogram transformation where the fragment's colour is replaced by another one with a look up table. This works well with a GLSL fragment shader but I doubt that it works with the SDL renderer on the GPU.

**Comment by HybridDog** [comment]
In case somebody want so try it out: I have created a demonstration program in a self-contained HTML file: https://hybriddog.github.io/stochastic_texture_sampling_demo/

There are small colour problems with the snow texture (see the violet patches) which also appear without a histogram transformation. I could not solve this; perhaps it is only solvable in the general case with an optimal transport solver or I have overlooked something.

![stsnow](https://user-images.githubusercontent.com/3192173/224548795-360ef00a-5f2b-48e0-a3a6-997bab0d5d92.png)
![2023-03-12-144339_1158x723_scrot](https://user-images.githubusercontent.com/3192173/224548787-b33b6dcc-b318-4874-9d7d-7331b7b3aa65.png)
